cou/apps/auxiliary.py

```diff
@@ -0,0 +1,165 @@
+
+    def _get_refresh_charm_plan(
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan for refreshing the current channel.
+
+        This function also identifies if charm comes from charmstore and in that case,
+        makes the migration.
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan for refreshing the charm.
+        :rtype: Optional[UpgradeStep]
+        """
+        switch = None
+        channel = self.channel
+
+        description = f"Refresh '{self.name}' to the latest revision of '{self.channel}'"
+
+        if self.charm_origin == "cs":
+            description = f"Migration of '{self.name}' from charmstore to charmhub"
+            switch = f"ch:{self.charm}"
+            channel = self.expected_current_channel
+        elif self.channel != self.expected_current_channel:
+            logger.warning(
+                "'%s' has the channel set to: %s which is different from the expected channel: %s",
+                self.name,
+                self.channel,
+                self.expected_current_channel,
+            )
+            return None
+
+        # pylint: disable=duplicate-code
+        return UpgradeStep(
+            description=description,
+            parallel=parallel,
+            function=async_upgrade_charm,
+            application_name=self.name,
+            channel=channel,
+            model_name=self.model_name,
+            switch=switch,
+        )
+
+    def upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Upgrade planning.
+
+        Auxiliary charms have multiple compatible OpenStack releases. In that case,
+        we check also the OpenStack origin from the "source" or "openstack-origin" to know
+        if it's necessary to change it. E.g: rabbitmq-server with workload version 3.8 is
+        compatible from ussuri to yoga. Even that is considered as yoga, we need to set the
+        source accordingly with the OpenStack components that might be in a version lower than
+        yoga.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation
+        :return: Plan that will add upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        os_origin_config = self.os_origin_config(target)
+        if self.current_os_release >= target and os_origin_config >= target:
+            msg = (
+                f"Application: '{self.name}' already configured for release equal or greater "
+                f"than {target}. Ignoring."
+            )
+            logger.info(msg)
+            raise HaltUpgradePlanGeneration(msg)
```

Comment:
I know you're aligning with the superclass (and that code is already merged), but two things occurred to me:
1. this function's role is only to produce a plan (if possible). Technically, it's not `upgrade_plan` that is choosing to ignore already upgraded apps, but rather the caller
2. similarly, an exception name should represent what happened, not what the `except` higher up in the stack may or may not do with it. This exception should rather be named `PlanGenerationError` if there's a genuine failure, or just return `None` if nothing needs to be done (and the plan generation - in a sense - succeeded)

Reply:
we can discuss that on the COU sync, but IMO the current approach makes sense.

If we return None in `upgrade_plan`, this means that `pre_upgrade_plan` and `post_upgrade_plan` can generate steps to execute and I think we don't want this.

I don't think this is an error from plan generation. It's just an application saying that for the target passed, we should do nothing to upgrade it.

Reply:
not really; whether `pre_upgrade_plan` and `post_upgrade_plan` steps will be generated (and executed) depends on how `generate_upgrade_plan` (`upgrade_plan`'s caller) will handle either `None` or an exception, which is my point: knowledge of the caller's implementation should not leak into your called method. I don't disagree with wanting to halt plan generation if we reach line 161 (I'd have to think more about it), I disagree with using an exception to signal a not-an-error situation.

`upgrade_plan` generates a plan to upgrade an application. If you ask an application to upgrade to a version it's already on, the plan is a no-op, and a no-op is not an error, so I think no exception should be raised.

[Here](https://stackoverflow.com/questions/1313812/raise-exception-vs-return-none-in-functions) is a pretty good discussion around the semantics of using returned objects vs raising exceptions to interact with callers.

Reply:
in any case, changes to these lines should be handled in a separate PR

Reply:
Created #55 to have a track of this issue.